src/main/java/com/google/devtools/build/lib/rules/repository/RepositoryDelegatorFunction.java

```diff
@@ -643,22 +648,27 @@ private String readMarkerFile(String content, Map<String, String> markerData) {
           firstLine = false;
         } else {
           int sChar = line.indexOf(' ');
-          String key = line;
-          String value = "";
           if (sChar > 0) {
-            key = unescape(line.substring(0, sChar));
-            value = unescape(line.substring(sChar + 1));
+            RepoRecordedInput input =
+                RepoRecordedInput.parse(unescape(line.substring(0, sChar)), baseDirectory);
+            if (input == null) {
+              // ignore invalid entries.
+              continue;
+            }
+            recordedInputValues.put(input, unescape(line.substring(sChar + 1)));
           }
-          markerData.put(key, value);
         }
       }
       return markerRuleKey;
     }
 
-    private String computeRuleKey(Rule rule) {
+    private String computeRuleKey(Rule rule, StarlarkSemantics starlarkSemantics) {
       return new Fingerprint()
           .addBytes(RuleFormatter.serializeRule(rule).build().toByteArray())
           .addInt(MARKER_FILE_VERSION)
+          // TODO: Using the hashCode() method for StarlarkSemantics here is technically wrong.
+          //   hashCode() is not guaranteed to be stable across JVM restarts.
```

Comment:
Only if using the default implementation of `hashCode()`, but `StarlarkSemantics` overrides it with logic that ultimately uses `hashCode()` on `ImmutableMap`, which is guaranteed to be stable (https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#hashCode--) as long as the `Object`s used as map values have stable hash code.

Reply:
But it's suboptimal for other reasons: It doesn't see changes to defaults. Could you add a comment about that?

Reply:
done. thanks I wasn't aware that Map had a documented hashcode algorithm (so does String, so it's actually completely stable).